version 1.0
## Copyright CMG@KIGM, Ales Maver

# WORKFLOW DEFINITION 
workflow CreateInterpretationTable {
  input {
    #File input_vcf
    #String? panel_gene_list

    String input_vcf="/mnt/EXOMES/PX5504/annotation/PX5504.annotated.vcf.gz"
    String? panel_gene_list = "ACE,ACTG2,AGT,AGTR1,ANOS1,BNC2,CHD7,DSTYK,EYA1,FRAS1,FREM1,FREM2,GATA3,GLI3,GRIP1,HAAO,HNF1B,HPSE2,ITGA8,KYNU,LRIG2,NPHP3,PAX2,PBX1,REN,RET,RPGRIP1L,RRM2B,SALL1,SIX5,TBX18,TRAP1,VPS33B,ACTA2,BICC1,BMP4,BSND,CHD1L,CHRM3,COX10,DACT1,DLG3,FOXC1,FOXC2,GDNF,GREM1,HCN3,KIT,MUC1,MYH11,NEK8,OCRL,ROBO2,SHH,SIX1,SLIT2,SMARCA4,SOX17,SPRY1,TNXB,TSHZ3,UMOD,UPK2,UPK3A,VIPAS39,AAAS,AARS2,AASS,ABAT,ABCA1,ABCB11,ABCB4,ABCB7,ABCC9,ABCD1,ABCD1,ABCD4,ABHD5,ACAD8,ACAD9,ACAD9,ACADM,ACADM,ACADS,ACADS,ACADSB,ACADVL,ACADVL,ACAT1,ACO2,ACO2,ACOX1,ACOX1,ACSF3,ACSL4,ACTA1,ACY1,ADAR,ADAR,ADK,ADNP,ADSL,AFF2,AFG3L2,AGA,AGK,AGL,AGL,AGPS,AGRN,AGXT,AHDC1,AIFM1,AIFM1,AKR1D1,AKT3,ALAD,ALDH18A1,ALDH18A1,ALDH4A1,ALDH4A1,ALDH5A1,ALDH5A1,ALDH6A1,ALDH7A1,ALDOA,ALDOA,ALDOB,ALG1,ALG1,ALG11,ALG11,ALG12,ALG12,ALG13,ALG3,ALG3,ALG6,ALG6,ALG8,ALG8,ALG9,AMACR,AMN,AMT,ANKRD11,ANO10,ANO5,AP1S1,AP1S2,AP3B1,AP3B2,AP4B1,AP4E1,AP4M1,AP4S1,APOA1,APOA5,APOC2,APOE,APOPT1,APOPT1,APRT,APTX,AR,ARCN1,ARG1,ARG1,ARHGEF6,ARHGEF9,ARID1A,ARID1B,ARID2,ARL13B,ARSA,ARSB,ARSE,ARV1,ARX,ASAH1,ASH1L,ASL,ASL,ASPA,ASPM,ASS1,ASS1,ASXL2,ASXL3,ATAD3A,ATIC,ATIC,ATL1,ATM,ATP13A2,ATP6AP1,ATP6AP2,ATP6V0A2,ATP6V0A2,ATP7A,ATP7A,ATP7B,ATP8B1,ATPAF2,ATRX,AUH,AUH,AUTS2,B3GALNT2,B3GALNT2,B3GALT6,B3GAT3,B3GLCT,B3GLCT,B4GALT1,B4GALT7,B4GAT1,BAAT,BAG3,BBS1,BBS10,BBS12,BBS2,BBS4,BBS5,BBS7,BBS9,BCKDHA,BCKDHB,BCKDK,BCKDK,BCL11A,BCOR,BCS1L,BICD2,BIN1,BMP4,BOLA3,BPTF,BRAF,BRF1,BRPF1,BRWD3,BSCL2,BTD,BUB1B,C12orf4,C12orf57,C12orf65,C19orf12,CA5A,CACNA1A,CACNA1D,CACNA1S,CAMK2A,CAMK2B,CAMTA1,CAPN3,CASK,CAT,CAV3,CBS,CBS,CC2D1A,CC2D2A,CCDC115,CCDC22,CCND2,CDC42,CDH11,CDH15,CDK13,CDK5RAP2,CDKL5,CENPF,CENPJ,CEP290,CEP41,CEP57,CEP83,CFL2,CHAT,CHCHD10,CHCHD10,CHD2,CHD4,CHD7,CHD8,CHKB,CHKB,CHMP1A,CHRNA1,CHRNB1,CHRND,CHRNE,CHRNG,CHST14,CHST3,CHST6,CHSY1,CKAP2L,CLCN4,CLN3,CLN3,CLN5,CLN6,CLN8,CLPB,CLPB,CLPP,CLTC,CNBP,CNKSR2,CNNM2,CNOT3,CNTNAP2,COASY,COG1,COG1,COG4,COG4,COG5,COG5,COG6,COG7,COG7,COG8,COG8,COL12A1,COL13A1,COL6A1,COL6A2,COL6A3,COLQ,COQ2,COQ4,COQ4,COQ6,COQ8A,COQ8A,COQ8B,COQ9,COX10,COX14,COX15,COX20,COX6A1,COX6B1,COX7B,CP,CPOX,CPS1,CPS1,CPT1A,CPT2,CPT2,CRADD,CREBBP,CRYAB,CSNK2A1,CSPP1,CTC1,CTCF,CTH,CTNNB1,CTNS,CTSA,CTSA,CTSD,CTSK,CUBN,CUL4B,CYB5R3,CYC1,CYP27A1,D2HGDH,DAG1,DAG1,DARS,DARS2,DARS2,DBH,DBT,DBT,DCHS1,DCX,DDC,DDOST,DDX3X,DEAF1,DES,DGUOK,DHCR24,DHCR7,DHFR,DHTKD1,DKC1,DLAT,DLD,DLG3,DMD,DMD,DNA2,DNAJB6,DNAJC12,DNAJC12,DNAJC19,DNAJC19,DNAJC5,DNM1,DNM1L,DNM2,DNMT3A,DOCK7,DOCK8,DOK7,DOLK,DOLK,DOLK,DPAGT1,DPAGT1,DPAGT1,DPM1,DPM2,DPM2,DPYD,DPYD,DPYS,DYM,DYNC1H1,DYNC1H1,DYRK1A,EARS2,EBF3,EBP,EBP,ECHS1,EED,EEF1A2,EHMT1,EIF2S3,EIF4A3,ELAC2,ELOVL4,ELP2,EMD,EML1,EMX2,ENO3,ENO3,EP300,EPG5,EPG5,EPM2A,ERCC2,ERCC3,ERCC5,ERCC6,ERCC8,ETFA,ETFA,ETFB,ETFB,ETFDH,ETFDH,ETHE1,ETHE1,EXOSC3,EXT1,EXT2,EXTL3,EZH2,FA2H,FAH,FAM126A,FAR1,FAR1,FARS2,FASTKD2,FBP1,FBXL4,FGD1,FGF12,FH,FHL1,FKBP14,FKRP,FKRP,FKRP,FKTN,FKTN,FKTN,FLAD1,FLNA,FLNC,FMN2,FMO3,FMR1,FOLR1,FOLR1,FOXG1,FOXP1,FOXP2,FOXRED1,FTSJ1,FUCA1,FUT8,G6PC,G6PC3,GAA,GAA,GABRG2,GALC,GALE,GALE,GALK1,GALNS,GALNT3,GALT,GAMT,GARS,GATM,GATM,GBA,GBE1,GBE1,GCDH,GCH1,GCLC,GDAP1,GDI1,GFAP,GFER,GFER,GFM1,GFM1,GFPT1,GFPT1,GIF,GK,GLA,GLB1,GLDC,GLI2,GLI3,GLIS3,GLRA1,GLRX5,GLUD1,GLUL,GLYCTK,GLYCTK,GM2A,GM2A,GMPPB,GMPPB,GMPPB,GNAI1,GNAS,GNB1,GNE,GNE,GNPAT,GNPAT,GNPTAB,GNPTAB,GNPTG,GNPTG,GNS,GPAA1,GPC3,GPD1,GPHN,GRHPR,GRIA3,GRIA4,GRID2,GRIK2,GRIN1,GRIN2A,GRIN2B,GRM1,GSS,GTPBP3,GTPBP3,GUSB,GUSB,GYG1,GYG1,GYS1,GYS1,GYS2,HAAO,HACE1,HADHA,HADHA,HADHB,HADHB,HAMP,HCCS,HCFC1,HCN1,HDAC4,HDAC8,HECW2,HEXA,HEXB,HEXB,HFE,HFE2,HGD,HGSNAT,HIBCH,HIST1H1E,HIVEP2,HLCS,HMBS,HMGCL,HMGCS2,HNRNPH2,HNRNPK,HNRNPU,HOGA1,HOXA1,HPD,HPRT1,HPRT1,HRAS,HSD17B10,HSD17B4,HSD3B7,HSPB1,HSPB8,HSPD1,HSPD1,HTRA2,HTRA2,HUWE1,HYAL1,IARS,IARS2,IBA57,IDH2,IDS,IDS,IDUA,IER3IP1,IER3IP1,IFIH1,IGF1,IGF1R,IGHMBP2,IL1RAPL1,INPP5K,INPP5K,IQSEC2,IRX5,ISCU,ISPD,ITGA7,ITPA,IVD,KANSL1,KARS,KAT6A,KBTBD13,KCNC1,KCNH1,KCNJ10,KCNJ11,KCNK9,KCNQ2,KCNQ5,KCTD7,KDM1A,KDM5B,KDM5C,KDM6A,KIAA0586,KIAA1109,KIDINS220,KIF11,KIF1A,KIF1BP,KIF5A,KIF5C,KIF7,KLHL40,KLHL41,KMT2A,KMT2C,KMT2D,KNL1,KPTN,KRAS,KYNU,L1CAM,L2HGDH,LAMA1,LAMA2,LAMA2,LAMC3,LAMP2,LAMP2,LAMP2,LARGE1,LARGE1,LARGE1,LARS2,LCAT,LCT,LDB3,LDHA,LDHA,LGI4,LIAS,LIG4,LINS1,LIPA,LIPT1,LMBRD1,LMNA,LMOD3,LONP1,LONP1,LPIN1,LPIN1,LPL,LRP2,LRP4,LRPPRC,MAGEL2,MAN1B1,MAN1B1,MAN2B1,MAN2B1,MANBA,MAOA,MAP2K2,MAP3K20,MARS2,MAT1A,MAT1A,MATR3,MBD5,MBOAT7,MBTPS2,MCCC1,MCCC1,MCCC2,MCCC2,MCEE,MCOLN1,MCOLN1,MCPH1,MDH2,MDH2,MECP2,MED12,MED13L,MED17,MEF2C,MEGF10,MFF,MFN2,MFSD8,MFSD8,MGAT2,MGAT2,MGME1,MICU1,MICU1,MID1,MKKS,MLYCD,MLYCD,MMAA,MMAB,MMACHC,MMADHC,MMADHC,MOCS2,MOGS,MOGS,MPDU1,MPI,MPI,MPV17,MRPS22,MSMO1,MTHFR,MTHFR,MTM1,MTO1,MTOR,MTR,MTR,MTRR,MTTP,MUSK,MUT,MYCN,MYH2,MYH3,MYH7,MYH8,MYMK,MYO18B,MYO5A,MYOT,MYT1L,NAA10,NAA15,NACC1,NAGA,NAGA,NAGLU,NAGS,NALCN,NARS2,NDP,NDST1,NDUFA1,NDUFA1,NDUFA10,NDUFA11,NDUFA2,NDUFAF1,NDUFAF2,NDUFAF3,NDUFAF4,NDUFAF5,NDUFAF6,NDUFB11,NDUFB3,NDUFS1,NDUFS1,NDUFS2,NDUFS3,NDUFS4,NDUFS6,NDUFS7,NDUFS8,NDUFV1,NDUFV2,NEB,NEU1,NEXMIF,NF1,NFIA,NFIX,NFU1,NGLY1,NHLRC1,NHS,NIPBL,NLGN3,NONO,NPC1,NPC1,NPC2,NPC2,NR2F1,NRXN1,NSD1,NSDHL,NSDHL,NSUN2,NT5C3A,NTRK1,NUBPL,OAT,OCRL,OFD1,OPA1,OPA3,OPA3,OPHN1,ORAI1,OTC,OTUD6B,OXCT1,PACS1,PAFAH1B1,PAH,PAH,PAK3,PARN,PAX6,PC,PCBD1,PCCA,PCCA,PCCB,PCCB,PCDH19,PCNT,PDE4D,PDGFRB,PDHA1,PDHB,PDHX,PDP1,PDSS1,PDSS1,PDSS2,PET100,PEX1,PEX10,PEX11B,PEX11B,PEX12,PEX13,PEX14,PEX16,PEX19,PEX2,PEX26,PEX3,PEX5,PEX6,PEX7,PEX7,PFKM,PFKM,PGAM2,PGAP1,PGAP2,PGAP3,PGAP3,PGK1,PGK1,PGK1,PGM1,PGM1,PGM3,PGM3,PHF6,PHF8,PHKA1,PHKA1,PHKA2,PHKB,PHKB,PHKG2,PHYH,PIGA,PIGL,PIGL,PIGN,PIGN,PIGO,PIGO,PIGT,PIGV,PIGV,PIK3R2,PLA2G6,PLA2G6,PLAA,PLEC,PLK4,PLP1,PLPBP,PMM2,PMM2,PMPCA,PNKP,PNPLA6,PNPO,PNPT1,POGZ,POLG,POLG,POLG2,POLG2,POMGNT1,POMGNT1,POMGNT1,POMGNT2,POMGNT2,POMK,POMT1,POMT1,POMT1,POMT2,POMT2,POMT2,PORCN,PPA2,PPM1D,PPOX,PPP2R5D,PPT1,PQBP1,PRKAG2,PRKAG2,PRKD1,PRODH,PRODH,PRPS1,PRSS12,PRUNE1,PSAP,PSMD12,PSPH,PTCH1,PTCHD1,PTDSS1,PTEN,PTPN11,PTPN23,PUF60,PURA,PUS1,PUS1,PYCR1,PYCR1,PYCR2,PYGL,PYGM,PYGM,QARS,QDPR,QRICH1,RAB11B,RAB39B,RAC1,RAD21,RAI1,RANBP2,RAPSN,RARB,RARS2,RBBP8,RBCK1,RBM10,RELN,RFT1,RFT1,RLIM,RMND1,RNASEH1,RNF125,ROBO3,RPGRIP1L,RPIA,RPL10,RPL10,RPS6KA3,RRM2B,RRM2B,RRM2B,RTTN,RYR1,SACS,SAMHD1,SAMHD1,SAR1B,SARS2,SATB2,SCN1A,SCN2A,SCN3A,SCN4A,SCN8A,SCO1,SCO2,SCP2,SDCCAG8,SDHA,SDHAF1,SDHB,SDHD,SEC23B,SELENON,SERAC1,SERAC1,SETBP1,SETD2,SETD5,SGCA,SGCB,SGCD,SGCG,SGSH,SGSH,SHANK2,SHANK3,SI,SIL1,SIL1,SLC12A6,SLC13A5,SLC16A1,SLC16A2,SLC17A5,SLC18A3,SLC19A2,SLC19A3,SLC22A5,SLC22A5,SLC25A1,SLC25A13,SLC25A15,SLC25A15,SLC25A19,SLC25A20,SLC25A22,SLC25A26,SLC25A3,SLC25A38,SLC25A4,SLC25A4,SLC25A46,SLC2A1,SLC2A2,SLC30A10,SLC33A1,SLC35C1,SLC35C1,SLC35D1,SLC37A4,SLC39A14,SLC39A14,SLC39A4,SLC39A8,SLC40A1,SLC46A1,SLC46A1,SLC4A4,SLC52A2,SLC52A2,SLC52A3,SLC52A3,SLC5A1,SLC5A7,SLC6A1,SLC6A17,SLC6A19,SLC6A20,SLC6A8,SLC6A9,SLC7A7,SLC9A6,SLX4,SMAD4,SMARCA2,SMARCA4,SMARCB1,SMARCE1,SMC1A,SMC3,SMN1,SMPD1,SMS,SNX14,SON,SOX10,SOX2,SOX3,SOX5,SPART,SPECC1L,SPEG,SPG11,SPG7,SPR,SPTAN1,SPTBN2,SQSTM1,SRCAP,SRD5A3,SRD5A3,SSR4,SSR4,ST3GAL3,ST3GAL5,ST3GAL5,STAC3,STIL,STIM1,STRA6,STX1B,STXBP1,SUCLA2,SUCLA2,SUCLG1,SUCLG1,SUMF1,SUOX,SURF1,SYN1,SYNE1,SYNGAP1,SYP,SZT2,TACO1,TAF1,TAF6,TALDO1,TANGO2,TAZ,TBC1D23,TBC1D24,TBCD,TBCE,TBL1XR1,TBR1,TCAP,TCF4,TECPR2,TFR2,TGIF1,TH,THOC2,THOC6,TIA1,TIMM8A,TK2,TK2,TLK2,TMCO1,TMEM165,TMEM165,TMEM216,TMEM240,TMEM5,TMEM5,TMEM67,TMEM70,TMEM70,TMTC3,TNNT1,TPK1,TPM2,TPM3,TPP1,TRAPPC11,TRAPPC9,TRIM32,TRIM37,TRIO,TRIP4,TRIT1,TRMT10A,TRMU,TRNT1,TRPM6,TRPV4,TSC1,TSC2,TSEN54,TSFM,TSFM,TSPAN7,TTC19,TTN,TUBA1A,TUBB2A,TUBB2B,TUBB3,TUSC3,TUSC3,TWIST1,TWNK,TYMP,TYMP,UBA1,UBA5,UBE2A,UBE3A,UBTF,UGT1A1,UMPS,UNC80,UPF3B,UQCRQ,UROC1,UROD,UROS,USP9X,VAMP1,VARS2,VCP,VLDLR,VMA21,VPS13B,VPS53,VRK1,WAC,WDR26,WDR45,WDR45B,WDR62,WDR73,WWOX,XDH,XYLT1,XYLT2,YARS2,YY1,ZBTB20,ZBTB24,ZDHHC9,ZEB2,ZFYVE26,ZIC2,ZMYND11,ZNF711,ZSWIM6,AARS,AASS,ABCD4,ABHD5,ACAD9,ACADM,ACTB,ACTG1,ACTL6A,ADAT3,ADGRG1,AFF4,AFG3L2,AGA,AGPS,AHI1,AIMP1,ALDH3A2,ALDH7A1,ALMS1,AMER1,AMPD2,AMT,ANKH,ARFGEF2,ARL6,ARMC9,ARSA,ARSB,ARSE,ASAH1,ASNS,ASPA,ASXL1,ATAD3A,ATP13A2,ATP1A1,ATP6V1A,ATP6V1B2,ATP8A2,ATR,B3GALNT2,B4GALNT1,B4GALT7,BCAP31,BCKDHA,BCKDHB,BCL11B,BCS1L,BLM,BOLA3,BRAT1,BTD,C12orf65,C1QBP,C5orf42,CA2,CA8,CACNA1C,CACNA1E,CACNA1G,CBL,CCBE1,CCDC47,CCDC88C,CDC6,CDK10,CDON,CEP135,CEP152,CHAMP1,CHD3,CIC,CIT,CLN5,CLN6,CLN8,CLP1,CNNM2,COG6,COL4A1,COL4A2,COL4A3BP,COLEC11,COX10,COX15,COX6B1,CRB2,CSNK2B,CSTB,CTDP1,CTNNA2,CTSD,CUX2,CWF19L1,CYC1,CYP2U1,D2HGDH,DARS,DCAF17,DDC,DDHD2,DDX11,DEPDC5,DHDDS,DHFR,DHTKD1,DHX30,DIAPH1,DIS3L2,DLD,DNMT3B,DOCK6,DPF2,DPM1,DPP6,DYM,DYSF,EFTUD2,EIF2AK3,EIF3F,ELAC2,EPG5,ERCC1,ERCC6L2,ERLIN2,ESCO2,ETFA,ETFB,ETFDH,EXOSC3,FAM111A,FAM20C,FAT4,FBXL4,FBXO11,FDXR,FH,FIG4,FLVCR1,FLVCR2,FOXRED1,FTCD,FUCA1,FUT8,GABRA1,GABRB2,GABRB3,GABRG2,GALC,GALT,GATAD2B,GCDH,GCH1,GJC2,GLB1,GLDC,GMPPA,GNAO1,GNS,GPSM2,GRIN2D,GTF2H5,GTPBP2,HADH,HADHA,HCCS,HCFC1,HEPACAM,HERC1,HESX1,HEXA,HGSNAT,HIBCH,HLCS,HMGCL,HSD17B4,HTRA2,HYLS1,IBA57,IDH2,IDUA,IFT172,IKBKG,INPP5E,INTS1,IRF2BPL,ISPD,ISPD,ITPR1,IVD,JAM3,KARS,KAT6B,KCNA2,KCNB1,KCNC3,KCNJ6,KCNQ3,KCNT1,L2HGDH,LAMB1,LARP7,LGI4,LRPPRC,LZTR1,MAB21L1,MAB21L2,MACF1,MAF,MAP2K1,MAPK8IP3,MASP1,MAST1,MCM3AP,MED23,MEIS2,METTL23,MFF,MGP,MKS1,MLC1,MMAA,MMAB,MMACHC,MOCS1,MPDU1,MPLKIP,MRPS22,MRPS34,MSL3,MSMO1,MTRR,MUT,MVK,NAGLU,NANS,NBEA,NDE1,NDUFS4,NDUFS7,NDUFS8,NDUFV1,NEU1,NFU1,NGLY1,NKX2-1,NPHP1,NRAS,NSD2,NT5C2,NTRK2,NUBPL,OCLN,ODC1,OGT,OSGEP,OTC,OTX2,PACS2,PAX8,PBX1,PC,PCGF2,PDHA1,PDHX,PDSS2,PEPD,PEX1,PEX10,PEX12,PEX13,PEX14,PEX16,PEX19,PEX2,PEX26,PEX3,PEX5,PEX6,PGAP2,PHGDH,PHIP,PIGA,PIGT,PIK3CA,PLCB1,POLG,POLR3A,POLR3B,POMGNT2,PPP1CB,PPP2R1A,PPT1,PRMT7,PSAP,PTF1A,PTS,QDPR,RAB18,RAB23,RAB3GAP1,RAB3GAP2,RAF1,RARS2,RBCK1,RERE,RHOBTB2,RIT1,RMND1,RNASEH2A,RNASEH2B,RNASEH2C,RNASET2,ROGDI,ROR2,RORA,RTEL1,RTN4IP1,SAMD9,SC5D,SCAPER,SCO1,SCO2,SDHA,SDHAF1,SET,SGPL1,SHANK1,SHH,SHOC2,SIK1,SIN3A,SIX3,SKI,SLC12A5,SLC17A5,SLC19A3,SLC1A2,SLC1A4,SLC25A1,SLC25A22,SLC35A1,SLC35A1,SLC35A2,SLC39A8,SLC6A19,SLC6A3,SMARCC2,SMOC1,SMPD1,SNAP29,SNRPB,SOS1,SOS2,SOX11,SOX4,SOX9,SPATA5,SPG11,SPRED1,STAG1,STAG2,STAMBP,STRADA,SUMF1,SUOX,SURF1,SYNJ1,SYT1,TANGO2,TAT,TAZ,TBC1D20,TBCK,TCN2,TCTN2,THRA,TMEM237,TMEM5,TOE1,TPP1,TREX1,TRIM32,TRIP12,TSEN2,TSEN34,TSEN54,TSFM,TSHB,TTC19,TTC37,TTC8,TTI2,TUBB,TUBB4A,TUBG1,TUBGCP6,UBE3B,UBR1,UMPS,VRK1,WASF1,WDPCP,WDR45,WDR81,XRCC4,YWHAG,ZBTB18,ZC4H2,ABAT,ABCB11,ABCC6,ABCG5,ABCG8,ABHD12,ACADVL,ACAN,ACAT1,ACP5,ACTA2,ACVR1,ACY1,ADA,ADA,ADCY5,ADGRG6,ADPRHL2,ADSL,AGL,AGO1,AGXT,AIMP2,AIPL1,AIRE,AK2,AKR1D1,AKT1,ALAD,ALAS2,ALDH3A2,ALDOA,ALDOB,ALG13,ALG2,ALPL,ALPL,ALS2,ALX3,ALX4,ANO5,ANTXR1,APOB,ARL14EP,ARMC4,ASAH1,ATP1A2,ATP1A3,ATP6V1B1,ATP7A,ATP8B1,B3GALT6,BFSP2,BGN,BHLHA9,BICD2,BMPER,BMPR1B,BSND,C19orf12,C2CD3,C2orf71,C4orf26,C8orf37,CARS2,CASQ1,CASR,CCDC103,CCDC114,CCDC40,CCDC65,CCDC78,CCNO,CCT5,CD96,CDH23,CDH3,CDK16,CDKN1C,CEP104,CEP63,CHL1,CHM,CHRDL1,CHRNG,CHST14,CHST3,CHSY1,CHUK,CIB2,CISD2,CLCN2,CLCN7,CLDN16,CLDN19,CLDN19,CNTN1,COA3,COA6,COL10A1,COL11A1,COL18A1,COL1A1,COL2A1,COL4A1,COL4A3,COL4A4,COL6A1,COL9A1,COL9A2,COL9A3,COMP,COQ9,COX7B,CRB1,CRBN,CRX,CRYAA,CRYBA1,CRYBB1,CRYBB2,CRYBB3,CRYGD,CSTB,CTNS,CTSC,CTSF,CTSK,CYCS,CYP1B1,CYP27A1,CYP7B1,CYP7B1,DCC,DCC,DCXR,DDB2,DDX53,DENND5A,DHCR24,DHCR7,DHDDS,DHODH,DIP2B,DLAT,DLG1,DLG2,DLG4,DLL3,DLL4,DMP1,DNAAF3,DNAAF4,DPM1,DPM3,DPM3,DPM3,DSPP,DSTYK,DVL1,DYNC2H1,ECEL1,EDA,EDNRA,EEF1B2,ELN,ENPP1,EOGT,EPB41L1,ERCC4,ERF,ERMARD,EVC,EVC2,EXOSC8,EXT2,EYA1,FA2H,FAAH2,FAM120C,FAM161A,FAM20A,FANCA,FANCC,FANCD2,FANCE,FANCF,FANCG,FANCI,FBN1,FBP1,FBXW4,FECH,FGF10,FGF14,FGF3,FGFR2,FGFR2,FHL1,FKBP14,FLNB,FLT4,FOXC1,FOXC2,FOXE1,FOXE3,FOXF1,FOXN1,FRAS1,FREM2,FRMPD4,FTCD,FTL,FTO,FXN,FXN,FYCO1,FZD6,GAA,GABBR2,GALK1,GAMT,GAS8,GATA2,GATA4,GATA6,GBA,GBA2,GDF5,GDF6,GHR,GJA1,GJA3,GJA8,GJB1,GJB2,GJB3,GK,GLE1,GLMN,GLUL,GNAI3,GNMT,GPR179,GRHL3,GRIA1,GRIA2,GRM6,GSS,GUCY2C,HACD1,HADH,HAX1,HINT1,HIST1H4C,HMGCS2,HNF1B,HNF4A,HNRNPDL,HOXA13,HOXC13,HOXD13,HPD,HPGD,HPS1,HPS1,HPSE2,HR,HSD17B10,HSD3B7,HSF4,HSPG2,HTT,HYAL1,HYDIN,IFITM5,IFT122,IFT43,IFT80,IGF2,IHH,IL11RA,IL1RAPL2,IMPAD1,INPPL1,IRF6,ISCU,ITGA7,JAG1,JAGN1,JAK3,KBTBD13,KCND3,KCNK4,KCNQ1,KCTD1,KIF22,KIF2A,KIRREL3,KIT,KLF1,KLHL15,KLHL40,KMT2B,KY,LBR,LDB3,LDLR,LDLRAP1,LEMD3,LFNG,LHX3,LHX4,LINGO1,LIPC,LMBRD1,LMNA,LMX1B,LNPK,LRP4,LRP5,LRRC6,LTBP2,LTBP3,MAGT1,MANBA,MAOA,MAP1B,MAP3K1,MAPK10,MAPRE2,MATN3,MC2R,MED25,MESP2,MFRP,MFSD2A,MIR17HG,MMP13,MNX1,MOCS1,MOCS2,MRPL3,MSX1,MSX2,MTFMT,MTMR14,MTO1,MTPAP,MVK,MYBPC3,MYH6,MYH8,MYH9,MYL1,MYO5B,MYO9A,MYPN,NAGS,NBAS,NBN,NDUFA4,NDUFAF2,NDUFB9,NECAP1,NEFL,NEK1,NFIB,NKX3-2,NMNAT1,NODAL,NOG,NOTCH2,NPHP3,NPHP4,NPHS1,NPR2,NR2F2,NR4A2,NR5A1,NRXN2,NT5C3A,NUDT2,NUP107,NUP62,NYX,OCRL,OPLAH,ORC6,OTOGL,OTULIN,PAK1,PANK2,PAPSS2,PARS2,PAX3,PAX9,PCK1,PCSK9,PDE6G,PDPR,PEPD,PET100,PGAM2,PGM1,PHF21A,PHGDH,PIGH,PIGM,PIK3R1,PINK1,PITX2,PITX3,PKD1L1,PKHD1,PLOD2,PMS2,PNP,PNPO,PNPT1,POC1B,POLD1,POLR1C,POLR1D,POR,POU1F1,PPA2,PRDM12,PROP1,PRPS1,PRRT2,PRSS56,PSAT1,PSMB8,PSPH,PTH1R,PTHLH,PTS,QARS,RASA1,RAX,RBP4,RBPJ,REEP1,RETREG1,RMRP,RNASEH2A,RNASEH2B,RNASEH2C,RNASET2,RNF135,RNU4ATAC,ROBO3,RPE65,RPGRIP1,RPS19,RPS23,RSPH1,RSPH3,RSPO4,RUNX2,RYR1,SACS,SALL1,SALL4,SBDS,SC5D,SCARF2,SCN11A,SCN4A,SDHAF2,SDHC,SEC23B,SETD1A,SETD1B,SETX,SETX,SH3PXD2B,SHROOM4,SIX1,SIX5,SKIV2L,SKIV2L,SLC12A3,SLC18A2,SLC25A12,SLC25A12,SLC25A26,SLC25A38,SLC26A2,SLC27A4,SLC2A1,SLC2A2,SLC35A2,SLC35A3,SLC35D1,SLC39A13,SLC3A1,SLC45A1,SLC4A1,SLC4A11,SLC5A5,SLC6A3,SLC6A8,SLC7A9,SMAD3,SMARCD2,SMCHD1,SMPD4,SNORD118,SOX17,SPAG1,SPEG,SPR,SPTLC1,SPTLC2,SRGAP3,SRY,ST3GAL3,STAR,STAT1,STS,STS,STT3A,STT3A,TAB2,TAF13,TARS2,TAT,TBC1D7,TBX1,TBX15,TBX20,TBX22,TBX3,TBX4,TBX5,TBXAS1,TCF12,TCF20,TCN2,TCTN3,TEK,TERT,TGFB1,TGFB2,TGFB3,TH,THRB,TIMM50,TMEM126B,TMEM126B,TMEM231,TMPRSS6,TNIK,TP63,TRAPPC11,TRAPPC2,TRAPPC6B,TREX1,TRIP11,TRMT1,TRPM1,TRPS1,TRPV4,TSHR,TTC37,TTPA,TUBA8,TUBGCP4,TUFM,TWIST2,TXNL4A,TYR,TYRP1,UGT1A1,UMOD,UPB1,UQCRB,UROC1,UROS,USB1,USP27X,UVSSA,VAMP1,VAPB,VIPAS39,VIPAS39,VKORC1,VPS33B,VPS33B,VPS33B,VSX2,WASHC5,WDR19,WDR34,WDR35,WDR4,WFS1,WNT1,WNT10B,WNT3,WNT7A,WRAP53,XPA,XPC,XPNPEP3,XYLT1,ZC3H14,ZFP57,ZIC1,ZIC3,ZMPSTE24,ZNF148,ZNF335,FMR1,AR,DMPK,DMPK,DMPK,FXN,ACTC1,ACVR2B,ALMS1,BCOR,BRAF,CBL,CHD7,CRELD1,ELN,FOXH1,GATA4,GATA6,GDF1,GJA1,GPC3,HAND1,HRAS,JAG1,KRAS,LEFTY2,MAP2K1,MAP2K2,MED13L,MEIS2,MYH6,NKX2-5,NKX2-6,NODAL,NOTCH1,NR2F2,NRAS,NSD1,PTPN11,RAF1,RIT1,SHOC2,SMAD6,SOS1,TBX1,TBX5,ZFPM2,ZIC3,AP1S2,ARID1A,ARID1B,ARL6,ARX,ASPM,ATRX,AVPR1A,BBIP1,BBS1,BBS2,BBS4,BBS5,BBS7,BBS9,BBS10,BBS12,BDNF,BRAF,CACNA1C,CASK,CBL,CDH8,CDKL5,CEP290,CHD7,CNTNAP2,CREBBP,DCX,DHCR7,DMD,EHMT1,ERCC6,ERCC8,EZH2,FGD1,FMR1,FOLR1,FOXG1,FOXP1,FOXP2,GABRB3,GLI3,GPC3,HD1C8,HOXA1,HPRT1,HRAS,KDM5C,KMT2D,KRAS,LICAM,LZTFL1,MAP2K1,MAP2K2,MBD5,MECP2,MED12,MEF2C,MET,MID1,MKKS,MKS1,NF1,NFIX,NHS,NIPBL,NLGN3,NLGN4X,NRAS,NRXN1,NSD1,OPHN1,PAFAH1B1,PCDH19,PHF6,PNKP,PQBP1,PTCH1,PTCHD1,PTEN,PTPN11,RAB39B,RAD21,RAF1,RAI1,RELN,RPGRIP1L,RPS6KA3,SCN1A,SDCCAG8,SHANK2,SHANK3,SHOC2,SHROOM4,SLC2A1,SLC6A4,SLC6A8,SLC9A6,SMARCA4,SMARCB1,SMARCE1,SMC1A,SMC3,SOS1,SPRED1,TCF4,TRAPPC9,TRIM32,TSC1,TSC2,TTC8,TUBA1A,UBE3A,VPS13B,WDPCP,ZEB2"

    String SnpEff_docker = "alesmaver/snpeff_v43:latest"
    String R_docker = "alesmaver/r-base"
  }  

  #File makeXSLSXoutputs_Rscript = "/home/ales/FastqToVCFPipeline/makeXLSXoutputs.R"
  String GenerateXLSXscriptUrl = "https://raw.githubusercontent.com/AlesMaver/CMGpipeline/master/makeXLSXoutputs.R"
  String sample_basename = basename(input_vcf, ".annotated.vcf.gz")

  # Get snpEff and dbNSFP annotations
    call GenerateVariantTable {
    input:
      input_vcf = input_vcf,
      panel_gene_list = panel_gene_list,
      sample_basename = sample_basename,
      docker = SnpEff_docker
    }

  # Download R script for XLSX file generation
    call GetGenerateXLSXscript {
    input:
      GenerateXLSXscriptUrl = GenerateXLSXscriptUrl,
      docker = SnpEff_docker
    }

  # Get snpEff and dbNSFP annotations
    call GenerateXLSX {
    input:
      RARE_FUNCTIONAL = GenerateVariantTable.RARE_FUNCTIONAL,
      HET_DOMINANT = GenerateVariantTable.HET_DOMINANT,
      COMPHET_RECESSIVE = GenerateVariantTable.COMPHET_RECESSIVE,
      HOM_RECESSIVE = GenerateVariantTable.HOM_RECESSIVE,
      CLINVAR_PATHOGENIC = GenerateVariantTable.CLINVAR_PATHOGENIC,
      CLINVAR_FILTERED = GenerateVariantTable.CLINVAR_FILTERED,
      CLINVAR_ALL = GenerateVariantTable.CLINVAR_ALL,
      PANEL_FILTERED = GenerateVariantTable.PANEL_FILTERED,
      PANEL_ALL = GenerateVariantTable.PANEL_ALL,

      sample_basename = sample_basename,

      makeXSLSXoutputs_Rscript = GetGenerateXLSXscript.makeXSLSXoutputs_Rscript,
      docker = R_docker
  }
 
  # Outputs that will be retained when execution is complete
  output {
    File XLSX_OUTPUT = GenerateXLSX.XLSX_OUTPUT
  }
}

##################
# TASK DEFINITIONS
##################
# Generate table of variants for interpretation
task GenerateVariantTable {
  input {
    # Command parameters
    File input_vcf
    String sample_basename
    String? panel_gene_list

    # Runtime parameters
    String docker
  }

  command {
    set -e
    SNPSIFT_EXTRACTFIELDS='/home/biodocker/bin/snpEff/scripts/vcfEffOnePerLine.pl | java -jar /home/biodocker/bin/snpEff/SnpSift.jar extractFields  - CHROM POS REF ALT GEN[0].GT GEN[0].AD GEN[0].DP GEN[0].GQ "ANN[*].GENE" "Disease_name" "Categorization" Inheritance Age HPO "OMIM" "ANN[*].FEATUREID" "ANN[*].HGVS_C" "ANN[*].RANK" "ANN[*].HGVS_P" "ANN[*].IMPACT" "ANN[*].EFFECT"  SLOpopulation.AC_Het SLOpopulation.AC_Hom SLOpopulation.AC_Hemi gnomAD.AC gnomAD.AF gnomAD.nhomalt gnomADexomes.AC gnomADexomes.AF gnomADexomes.nhomalt clinvar.CLNSIG clinvar.CLNDN clinvar.CLNHGVS clinvar.CLNSIGCONF clinvar.CLNSIGINCL dbNSFP_REVEL_rankscore  dbNSFP_MetaSVM_pred dbNSFP_CADD_phred  dbNSFP_DANN_rankscore dbNSFP_SIFT_pred  dbNSFP_SIFT4G_pred  dbNSFP_Polyphen2_HDIV_pred  dbNSFP_MutationTaster_pred dbNSFP_PrimateAI_pred dbNSFP_Polyphen2_HDIV_score SpliceAI.SpliceAI dbscSNV.ada_score dbscSNV.rf_score dbNSFP_GERP___NR  dbNSFP_GERP___RS dbNSFP_Interpro_domain pLI oe_mis pRec "LOF[*].GENE" "LOF[*].GENEID" "LOF[*].NUMTR" "LOF[*].PERC" "NMD[*].GENE" "NMD[*].GENEID" "NMD[*].NUMTR" "NMD[*].PERC"'

    # Optional fields available in the VCF files, consider adding them later
    # dbNSFP_GTEx_V7_gene dbNSFP_GTEx_V7_tissue dbNSFP_Geuvadis_eQTL_target_gene dbNSFP_Aloft_Fraction_transcripts_affected  dbNSFP_Aloft_prob_Tolerant  dbNSFP_Aloft_prob_Recessive  dbNSFP_Aloft_prob_Dominant  dbNSFP_Aloft_pred dbNSFP_Aloft_Confidence 
    zcat ~{input_vcf} | java -jar /home/biodocker/bin/snpEff/SnpSift.jar filter "((( gnomAD.AF < 0.05 ) | !( exists gnomAD.AC )) & (( gnomAD.nhomalt < 10 ) | !( exists gnomAD.nhomalt )) & (( SLOpopulation.AC_Het < 100 ) | !( exists SLOpopulation.AC_Het )) & (!( exists SLOpopulation.AC_Hom ) | ( SLOpopulation.AC_Hom <= 6 ))) & (( ANN[*].IMPACT has 'MODERATE') | (ANN[*].IMPACT has 'HIGH') | (dbscSNV.ada_score > 0.5) | (dbscSNV.rf_score > 0.5))" | eval $SNPSIFT_EXTRACTFIELDS > ~{sample_basename}.RARE_FUNCTIONAL.tab

    zcat ~{input_vcf} | java -jar /home/biodocker/bin/snpEff/SnpSift.jar filter "((( gnomAD.AC < 10 ) | !( exists gnomAD.AC )) & (( SLOpopulation.AC_Het < 5 ) | !( exists SLOpopulation.AC_Het )) & (!( exists SLOpopulation.AC_Hom ) | ( SLOpopulation.AC_Hom <= 6 ))) & (( ANN[*].IMPACT has 'MODERATE') | (ANN[*].IMPACT has 'HIGH') | (dbscSNV.ada_score > 0.5) | (dbscSNV.rf_score > 0.5)) & (( Inheritance =~ '.*AD.*' ) | ( HPO =~ '.*Autosomal_dominant.*' ) | ( HPO  =~  '.*X-linked_dominant.*' )) & isHet(GEN[0].GT)" | eval $SNPSIFT_EXTRACTFIELDS > ~{sample_basename}.HET_DOMINANT.tab

    zcat ~{input_vcf} | java -jar /home/biodocker/bin/snpEff/SnpSift.jar filter "((( gnomAD.AF < 0.05 ) | !( exists gnomAD.AC )) & (( gnomAD.nhomalt < 10 ) | !( exists gnomAD.nhomalt )) & (( SLOpopulation.AC_Het < 100 ) | !( exists SLOpopulation.AC_Het )) & (!( exists SLOpopulation.AC_Hom ) | ( SLOpopulation.AC_Hom <= 6 ))) & (( ANN[*].IMPACT has 'MODERATE') | (ANN[*].IMPACT has 'HIGH') | (dbscSNV.ada_score > 0.5) | (dbscSNV.rf_score > 0.5)) & (( Inheritance =~ '.*AR.*' ) | ( Inheritance =~ '.*XL.*' ) | ( HPO =~ '.*Autosomal_recessive.*' ) | ( HPO  =~  '.*X-linked_recessive.*' )) & !(isHom(GEN[0].GT))" | eval $SNPSIFT_EXTRACTFIELDS >  ~{sample_basename}.COMPHET_RECESSIVE.tab

    zcat ~{input_vcf} | java -jar /home/biodocker/bin/snpEff/SnpSift.jar filter "((( gnomAD.AF < 0.05 ) | !( exists gnomAD.AC )) & (( gnomAD.nhomalt < 10 ) | !( exists gnomAD.nhomalt )) & (( SLOpopulation.AC_Het < 100 ) | !( exists SLOpopulation.AC_Het )) & (!( exists SLOpopulation.AC_Hom ) | ( SLOpopulation.AC_Hom <= 6 ))) & (( ANN[*].IMPACT has 'MODERATE') | (ANN[*].IMPACT has 'HIGH') | (dbscSNV.ada_score > 0.5) | (dbscSNV.rf_score > 0.5)) & (( Inheritance =~ '.*AR.*' ) | ( Inheritance =~ '.*XL.*' ) | ( HPO =~ '.*Autosomal_recessive.*' ) | ( HPO  =~  '.*X-linked_recessive.*' )) & isHom(GEN[0].GT)" | eval $SNPSIFT_EXTRACTFIELDS > ~{sample_basename}.HOM_RECESSIVE.tab

    zcat ~{input_vcf} | java -jar /home/biodocker/bin/snpEff/SnpSift.jar filter "((( gnomAD.AF < 0.05 ) | !( exists gnomAD.AC )) & (( gnomAD.nhomalt < 10 ) | !( exists gnomAD.nhomalt )) & (( SLOpopulation.AC_Het < 100 ) | !( exists SLOpopulation.AC_Het )) & (!( exists SLOpopulation.AC_Hom ) | ( SLOpopulation.AC_Hom <= 6 ))) & (( ANN[*].IMPACT has 'MODERATE') | (ANN[*].IMPACT has 'HIGH') | (dbscSNV.ada_score > 0.5) | (dbscSNV.rf_score > 0.5)) & ((clinvar.CLNSIGCONF =~ '.*Pathogenic.*') | (clinvar.CLNSIGCONF =~ '.*Likely_pathogenic.*') | (clinvar.CLNSIG =~ '.*Likely_pathogenic.*') | (clinvar.CLNSIG =~ '.*Pathogenic.*'))" | eval $SNPSIFT_EXTRACTFIELDS > ~{sample_basename}.CLINVAR_PATHOGENIC.tab

    zcat ~{input_vcf} | java -jar /home/biodocker/bin/snpEff/SnpSift.jar filter "((( gnomAD.AF < 0.05 ) | !( exists gnomAD.AC )) & (( gnomAD.nhomalt < 10 ) | !( exists gnomAD.nhomalt )) & (( SLOpopulation.AC_Het < 100 ) | !( exists SLOpopulation.AC_Het )) & (!( exists SLOpopulation.AC_Hom ) | ( SLOpopulation.AC_Hom <= 6 ))) & (( ANN[*].IMPACT has 'MODERATE') | (ANN[*].IMPACT has 'HIGH') | (dbscSNV.ada_score > 0.5) | (dbscSNV.rf_score > 0.5)) & (exists clinvar.CLNSIG)" | eval $SNPSIFT_EXTRACTFIELDS > ~{sample_basename}.CLINVAR_FILTERED.tab

    zcat ~{input_vcf} | java -jar /home/biodocker/bin/snpEff/SnpSift.jar filter "(exists clinvar.CLNSIG) & !((clinvar.CLNSIG =~ '.*Likely_benign.*') | (clinvar.CLNSIG =~ '.*Benign.*'))" | eval $SNPSIFT_EXTRACTFIELDS > ~{sample_basename}.CLINVAR_ALL.tab

    ~{if defined(panel_gene_list) then "echo " + panel_gene_list + " | awk '{gsub(/\,/,\"\\n\",$0);print $0}' > panel_gene_list.txt" else ""}

    if [ -f panel_gene_list.txt ]; then
      zcat ~{input_vcf} | java -jar /home/biodocker/bin/snpEff/SnpSift.jar filter -s panel_gene_list.txt "((( gnomAD.AF < 0.05 ) | !( exists gnomAD.AC )) & (( gnomAD.nhomalt < 10 ) | !( exists gnomAD.nhomalt )) & (( SLOpopulation.AC_Het < 100 ) | !( exists SLOpopulation.AC_Het )) & (!( exists SLOpopulation.AC_Hom ) | ( SLOpopulation.AC_Hom <= 6 ))) & (( ANN[*].IMPACT has 'MODERATE') | (ANN[*].IMPACT has 'HIGH') | (dbscSNV.ada_score > 0.5) | (dbscSNV.rf_score > 0.5)) & (ANN[*].GENE in SET[0])" | eval $SNPSIFT_EXTRACTFIELDS > ~{sample_basename}.PANEL_FILTERED.tab

      zcat ~{input_vcf} | java -jar /home/biodocker/bin/snpEff/SnpSift.jar filter -s panel_gene_list.txt "ANN[*].GENE in SET[0]" | eval $SNPSIFT_EXTRACTFIELDS > ~{sample_basename}.PANEL_ALL.tab
    fi


  }
  runtime {
    docker: docker
  }
  output {
    File RARE_FUNCTIONAL = "~{sample_basename}.RARE_FUNCTIONAL.tab"
    File HET_DOMINANT = "~{sample_basename}.HET_DOMINANT.tab"
    File COMPHET_RECESSIVE = "~{sample_basename}.COMPHET_RECESSIVE.tab"
    File HOM_RECESSIVE = "~{sample_basename}.HOM_RECESSIVE.tab"
    File CLINVAR_PATHOGENIC = "~{sample_basename}.CLINVAR_PATHOGENIC.tab"
    File CLINVAR_FILTERED = "~{sample_basename}.CLINVAR_FILTERED.tab"
    File CLINVAR_ALL = "~{sample_basename}.CLINVAR_ALL.tab"
    File? PANEL_FILTERED = "~{sample_basename}.PANEL_FILTERED.tab"
    File? PANEL_ALL = "~{sample_basename}.PANEL_ALL.tab"
  }
}

task GetGenerateXLSXscript {
  input {
    String GenerateXLSXscriptUrl

    # Runtime parameters
    String docker
  }

  command {
    set -e
    wget ~{GenerateXLSXscriptUrl}
  }

  runtime {
    docker: docker
  }

  output {
    File makeXSLSXoutputs_Rscript = "makeXLSXoutputs.R"
  }
}

# Generate table of variants for interpretation
task GenerateXLSX {
  input {
    # Command parameters
    File RARE_FUNCTIONAL
    File HET_DOMINANT
    File COMPHET_RECESSIVE
    File HOM_RECESSIVE
    File CLINVAR_PATHOGENIC
    File CLINVAR_FILTERED
    File CLINVAR_ALL
    File? PANEL_FILTERED
    File? PANEL_ALL
    String sample_basename

    File makeXSLSXoutputs_Rscript

    # Runtime parameters
    String docker
  }

  command {
  Rscript ~{makeXSLSXoutputs_Rscript} --sample_basename=~{sample_basename} --RARE_FUNCTIONAL=~{RARE_FUNCTIONAL} --HET_DOMINANT=~{HET_DOMINANT} --COMPHET_RECESSIVE=~{COMPHET_RECESSIVE} --HOM_RECESSIVE=~{HOM_RECESSIVE} --CLINVAR_PATHOGENIC=~{CLINVAR_PATHOGENIC} --CLINVAR_FILTERED=~{CLINVAR_FILTERED} --CLINVAR_ALL=~{CLINVAR_ALL} ~{if defined(PANEL_FILTERED) then " --PANEL_FILTERED " + PANEL_FILTERED else ""}  ~{if defined(PANEL_ALL) then " --PANEL_ALL " + PANEL_ALL else ""} --XLSX_OUTPUT=~{sample_basename}.FinalReportNew.xlsx 
    
  }
  runtime {
    docker: docker
  }
  output {
    File XLSX_OUTPUT = "~{sample_basename}.FinalReportNew.xlsx"

  }
}

